profiling/src/profile/mod.rs

```diff
@@ -527,15 +535,21 @@ impl Profile {
 
         let id = match self.samples.get_index_of(&s) {
             None => {
-                self.samples.insert(s, values);
+                let ts_map = HashMap::from([(timestamp, values)]);
+                self.samples.insert(s, ts_map);
                 PProfId(self.samples.len())
             }
             Some(index) => {
                 let (_, existing_values) =
                     self.samples.get_index_mut(index).expect("index to exist");
-                for (a, b) in existing_values.iter_mut().zip(values) {
-                    a.add_assign(b)
+                if let Some(existing_values) = existing_values.get_mut(&timestamp) {
+                    for (a, b) in existing_values.iter_mut().zip(values) {
+                        a.add_assign(b)
+                    }
+                } else {
+                    existing_values.insert(timestamp, values);
```

Comment:
I would think that most of the time we will go through l.550 but I might be wrong, which means that we would trade CPU against memory.